quadlet install: add support for multiple quadlets in a single file

flouthoc opened the pull request:
Add ability to `quadlet install` to install multiple quadlets from a single file by separating
them with `---` delimiters, similar to how multiple quadlets can be
installed from a directory. Each section requires `# FileName=<name>` comment for custom naming.

Example usage:
```
  # FileName=web-server
  [Container]
  Image=nginx:latest
  PublishPort=8080:80
  
  ---
  
  # FileName=app-storage
  [Volume]
  Label=app=webapp
  ```
  ```
  $ podman quadlet install webapp.quadlets
  /home/user/.config/containers/systemd/web-server.container
  /home/user/.config/containers/systemd/app-storage.volume
```

<!--
Thanks for sending a pull request!

Please make sure you've read our contributing guidelines and how to submit a pull request (https://github.com/containers/podman/blob/main/CONTRIBUTING.md#submitting-pull-requests).

Finally, be sure to sign commits with your real name. Since by opening
a PR you already have commits, you can add signatures if needed with
something like `git commit -s --amend`.
-->

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes, please follow the Kubernetes model:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
quadlet install: add support for multiple quadlets in a single file
```

flouthoc committed:
quadlet: add support for multiple quadlets in a single file

Enable installing multiple quadlets from one file using '---' delimiters.
Each section requires '# FileName=<name>' comment for custom naming.
Single quadlet files remain unchanged for backward compatibility.

Assited by: claude-4-sonnet

Signed-off-by: flouthoc <flouthoc.git@gmail.com>

Luap99 commented:
@ygalblum there was an internal design doc https://docs.google.com/document/d/1fVjmxVA6k_83iVztQ4j4ISGW7JMVb7yCBfO6vYk9SWs/edit?usp=sharing

If that wasn't shared to you then this is already a big issue as you as main quadlet maintainer should be able to look at these things and give your input. As I noted there design docs shouldn't be internal at all and be public in this repo so all maintainers and community people can actually give feedback https://github.com/containers/podman/tree/main/contrib/design-docs which was also not done here.

> So, as you, I assume this was decided somewhere and just want to make sure we keep the distinction.

Well the way I see it decided "somewhere" doesn't count in general, we decide upstream on github what gets merged which includes all maintainers including you of course. So if you think this is a bad idea you can still speak up. I already objected that feature internally and on the issue (https://github.com/containers/podman/issues/26447) but it seemed like some other maintainers on the team were in favour and others abstained so I didn't push any further against this.

Honny1 reviewed (commented):
My only thought is that we could improve performance by reading the input file only once instead of multiple times.

ygalblum commented:
I don't have access to this doc either. I remember the discussion we had on Quadlet's support for single file and I agree that Quadlet should not support it.

Generally speaking about `podman quadlet`, my personal opinion is that I don't see its benefit in Linux systems where users can simply copy the files and use `systemctl` to manage the services. I can see it in MacOS or Windows where the details of where podman actually runs are hidden from the user and copying the files manually is not obvious.

To me, the `podman quadlet` command is more of a `copy` tool than actually Quadlet and that's how I try to differentiate between them. I don't mind having a tool that knows to read a single file and break it into many. But, it should have a distinct extension (one not supported by Quadlet), schema and proper documentation.

eriksjolund commented:
> There is no reason why this should be private IMO. I don't own the doc so I cannot share it. As I said for feedback and public knowledge this should have been posted here on github into the design-doc directory instead of being a private doc.
> 
> There have been a ton of design docs internally over the years. I am trying to push all to do these discussions in public here on github for transparency and to allow for feedback outside our team early on in the design phase.

That is a good idea.

flouthoc commented:
> I don't have access to this doc either. I remember the discussion we had on Quadlet's support for single file and I agree that Quadlet should not support it.
> 
> Generally speaking about `podman quadlet`, my personal opinion is that I don't see its benefit in Linux systems where users can simply copy the files and use `systemctl` to manage the services. I can see it in MacOS or Windows where the details of where podman actually runs are hidden from the user and copying the files manually is not obvious.
> 
> To me, the `podman quadlet` command is more of a `copy` tool than actually Quadlet and that's how I try to differentiate between them. I don't mind having a tool that knows to read a single file and break it into many. But, it should have a distinct extension (one not supported by Quadlet), schema and proper documentation.

@ygalblum You should have access to design doc now.

>  But, it should have a distinct extension (one not supported by Quadlet), schema and proper documentation

I agree, I think we can limit this to one particular extension maybe like `.quadlets`. If everyone agrees I can make this change. WDYT ?

flouthoc committed:
quadlet install: multiple quadlets from single file should share app

Quadlets installed from `.quadlet` file now belongs to a single
application, anyone file removed from this application removes all the
other files as well.

Assited by: claude-4-sonnet

Signed-off-by: flouthoc <flouthoc.git@gmail.com>

flouthoc commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -325,6 +403,172 @@ func appendStringToFile(filePath, text string) error {
+// parseMultiQuadletFile parses a file that may contain multiple quadlets separated by "---"
+// Returns a slice of quadletSection structs, each representing a separate quadlet
+func parseMultiQuadletFile(filePath string) ([]quadletSection, error) {
+	content, err := os.ReadFile(filePath)
```
This function is removed.

flouthoc commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -209,13 +217,84 @@ func (ic *ContainerEngine) QuadletInstall(ctx context.Context, pathsOrURLs []str
+			if isQuadletsFile {
+				// For .quadlets files, always treat as multi-quadlet (even single quadlets)
+				isMulti = true
+			}
```
Done.

Honny1 reviewed (commented):
LGTM

TomSweeneyRedHat commented:
Code LGTM, and I really like making this into a new file type so the old quadlet files are handled as before.  I do have a couple of smal tweaks for the doc that I'd like to see happen, and then I'd like a head nod from @ygalblum

TomSweeneyRedHat commented on docs/source/markdown/podman-quadlet-install.1.md:
```diff
@@ -16,7 +16,9 @@ This command allows you to:
-Note: If a quadlet is part of an application, removing that specific quadlet will remove the entire application. When a quadlet is installed from a directory, all files installed from that directory—including both quadlet and non-quadlet files—are considered part of a single application.
+    * Install multiple Quadlets from a single file with `.quadlets` extension where each Quadlet is separated by a `---` delimiter. When using multiple quadlets in a single file, each quadlet section must include a `# FileName=<name>` comment to specify the name for that quadlet.
+
+Note: If a quadlet is part of an application, removing that specific quadlet will remove the entire application. When a quadlet is installed from a directory, all files installed from that directory—including both quadlet and non-quadlet files—are considered part of a single application. Similarly, when multiple quadlets are installed from a single `.quadlets` file, they are all considered part o ...
```
If you have a bunch of quadlet files in a directory, and also one or more `.quadlets` files, are all of them processed if you tell the system to use everything in the directory?  Or if you have multiple `.quadlets` files in a directory and you specify the directory, will they all be installed?

Are any of these scenarios things you need to test here?

flouthoc commented:
> Code LGTM, and I really like making this into a new file type so the old quadlet files are handled as before. I do have a couple of smal tweaks for the doc that I'd like to see happen, and then I'd like a head nod from @ygalblum

Done. addressed comments. @ygalblum Could you PTAL

ygalblum commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -209,13 +217,77 @@ func (ic *ContainerEngine) QuadletInstall(ctx context.Context, pathsOrURLs []str
+				if assetFile != "" {
+					// This is part of an app installation, allow non-quadlet files as assets
+					// Don't validate as quadlet file (validateQuadletFile will be false)
+				} else {
```
Don't add empty contexts
```suggestion
                // Allow non-quadlet files as assets
				// Don't validate as quadlet file (validateQuadletFile will be false)
				if assetFile == "" {
```

ygalblum commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -209,13 +217,77 @@ func (ic *ContainerEngine) QuadletInstall(ctx context.Context, pathsOrURLs []str
+
+					// Write the quadlet content to the temporary file
+					_, err = tmpFile.WriteString(quadlet.content)
+					if err != nil {
```
Close the file once regardless to success
```suggestion
					_, err = tmpFile.WriteString(quadlet.content)
					tmpFile.Close()
					if err != nil {
```

ygalblum commented on test/system/254-podman-quadlet-multi.bats:
```diff
@@ -0,0 +1,371 @@
+#!/usr/bin/env bats   -*- bats -*-
```
None of the test use additional systemd sections. Need to make sure that these section both don't interfere with the creation of the standalone files and that they are copied correctly

flouthoc replied:
Done, added in first test.

ygalblum reviewed (commented):
Sorry for the delay in my review

flouthoc commented on test/system/254-podman-quadlet-multi.bats:
```diff
@@ -0,0 +1,371 @@
+    assert "$output" =~ "missing required.*FileName" "error should mention missing FileName"
+}
+
+@test "quadlet verb - multi-quadlet file creates application" {
```
Done.

flouthoc commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -325,6 +405,154 @@ func appendStringToFile(filePath, text string) error {
+
+		// Extract name for this quadlet section
+		var name string
+		if isMultiSection {
```
This was explicitly decided in a design meeting and design doc, to defer naming of file to the user.

flouthoc commented:
> So if there's only one section, FileName is not required?

Yes it is only needed to differentiate naming of multiple sections, not required if there is only one section.

ygalblum commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -209,13 +217,67 @@ func (ic *ContainerEngine) QuadletInstall(ctx context.Context, pathsOrURLs []str
+					}
+
+					// Record the installation (use a unique key for each section)
+					sectionKey := fmt.Sprintf("%s#%s", toInstall, quadlet.name)
```
But in this code section, since you are using only the `quadlet.name` you are overriding what was set in the map. Is that OK?

ygalblum commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -209,13 +217,65 @@ func (ic *ContainerEngine) QuadletInstall(ctx context.Context, pathsOrURLs []str
+					destName := quadlet.name + quadlet.extension
+					installedPath, err := ic.installQuadlet(ctx, tmpFile.Name(), destName, installDir, assetFile, true, options.Replace)
+					if err != nil {
+						installReport.QuadletErrors[toInstall] = fmt.Errorf("unable to install quadlet section %s: %w", quadlet.name, err)
```
```suggestion
						installReport.QuadletErrors[toInstall] = fmt.Errorf("unable to install quadlet section %s: %w", destName, err)
```

ygalblum commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -209,13 +217,65 @@ func (ic *ContainerEngine) QuadletInstall(ctx context.Context, pathsOrURLs []str
+					}
+
+					// Record the installation (use a unique key for each section)
+					sectionKey := fmt.Sprintf("%s#%s", toInstall, quadlet.name+quadlet.extension)
```
```suggestion
					sectionKey := fmt.Sprintf("%s#%s", toInstall, destName)
```

ygalblum commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -325,6 +393,125 @@ func appendStringToFile(filePath, text string) error {
+				if fileName == "" {
+					return "", fmt.Errorf("FileName comment found but no filename specified")
+				}
+				// Validate filename (basic validation - no path separators, no extensions)
```
You're not validating no extensions anymore
```suggestion
				// Validate filename (basic validation - no path separators)
```

ygalblum reviewed (commented):
Minor nits

ygalblum reviewed (approved):
LGTM

Honny1 commented:
LGTM

cc @TomSweeneyRedHat

TomSweeneyRedHat commented:
/lgtm

leuasseurfarrelds247-arch reviewed (changes requested):
e787b4f5031ee76b92b66c1e210370ae112599fa
